src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1704 @@
+				// We planned for more items than we may take. Can happen when
+				// transfers are cancelled in the last moment. Ignore.
+				break;
+			}
+		}
+	}
+
+	plan->second.pop_front();
+	if (plan_size > 1) {
+		ship.set_destination(game, plan->second.front().dock.get(game));
+		sslog("Loaded cargo and sending to %u\n\n", plan->second.front().dock.serial());
+	} else {
+		ship.set_destination(game, nullptr);
+		assert(ship.get_nritems() == 0);
+		sslog("Loaded nothing, idle now\n\n");
+	}
+}
+
+// `dock` is not a dangling reference yet, but this function is called
+// via `ShipFleet::remove_port()` from `PortDock::cleanup()`
+void ShippingSchedule::port_removed(Game& game, PortDock* dock) {
+	sslog("\nShippingSchedule::port_removed (%u)\n", dock->serial());
+	// Find all ships planning to visit this dock and reroute them.
+	std::vector<Ship*> ships_heading_there;
+	for (auto& pair : plans_) {
+		size_t index_of_deleted_dock = 0;
+		const size_t nr_entries = pair.second.size();
+		bool dock_found = false;
+		for (auto& ss : pair.second) {
+			if (ss.dock == dock) {
+				dock_found = true;
+			}
+			auto it = ss.load_there.find(dock);
+			if (it != ss.load_there.end()) {
+				assert(!dock_found);
+				ss.load_there.erase(it);
+				// TODO(Nordfriese): Perhaps the ship could not skip some destinations
+				// (like in the third pass)
+			}
+			if (!dock_found) {
+				++index_of_deleted_dock;
+			}
+		}
+		if (!dock_found) {
+			continue;
+		}
+		ships_heading_there.push_back(pair.first.get(game));
+		if (index_of_deleted_dock == 0) {
+			// reroute to next dock
+			pair.second.pop_front();
+			if (pair.second.empty()) {
+				// no other docks to visit, but wares left, reroute to the closest one
+				if (fleet_.get_ports().empty() || pair.first.get(game)->get_nritems() == 0) {
+					// PANIC! There are no ports at all left!!
+					// But we still have cargo!!! What should we do????
+					// Stay calm. Just do nothing. Nothing at all.
+					log("Ship %s is carrying %u items OR there are no ports left, setting NO "
+					    "destination\n",
+					    pair.first.get(game)->get_shipname().c_str(),
+					    pair.first.get(game)->get_nritems());
+					pair.first.get(game)->set_destination(game, nullptr);
+				} else {
+					PortDock* closest = nullptr;
+					int32_t dist = 0;
+					for (PortDock* pd : fleet_.get_ports()) {
+						Path path;
+						int32_t d = -1;
+						pair.first.get(game)->calculate_sea_route(game, *pd, &path);
+						game.map().calc_cost(path, &d, nullptr);
+						assert(d >= 0);
+						if (!closest || d < dist) {
+							dist = d;
+							closest = pd;
+						}
+					}
+					assert(closest);
+					log("Ship %s is carrying %u items, rerouting to NEW destination %u\n",
+					    pair.first.get(game)->get_shipname().c_str(),
+					    pair.first.get(game)->get_nritems(), closest->serial());
+					pair.second.push_back(SchedulingState(closest, false, dist));
+					pair.first.get(game)->set_destination(game, closest);
+				}
+			} else {
```

Comment:
Loosing track about the else here, we need a comment or recftor this into smallep methods.

MARK for ME -- continue review here